worker/solver_worker.ml

```diff
@@ -1,102 +1,171 @@
-        | ex -> Lwt.return (Fmt.error_msg "%a" Fmt.exn ex))
-end
+  let jobs_accepted =
+    let help = "Number of jobs accepted in total" in
+    Counter.v ~help ~namespace ~subsystem "jobs_accepted_total"
 
-let solve_to_custom req =
-  let open Cluster_api.Raw in
-  let params =
-    Yojson.Safe.to_string
-    @@ Solver_service_api.Worker.Solve_request.to_yojson req
-  in
-  let custom = Builder.Custom.init_root () in
-  let builder = Builder.Custom.payload_get custom in
-  let request =
-    Solver_service_api.Raw.Builder.Solver.Solve.Params.init_pointer builder
-  in
-  Solver_service_api.Raw.Builder.Solver.Solve.Params.request_set request params;
-  let r = Reader.Custom.of_builder custom in
-  Reader.Custom.payload_get r
+  let job_time =
+    let help = "Time jobs ran for" in
+    Summary.v_label ~label_name:"result" ~help ~namespace ~subsystem
+      "job_time_seconds"
+
+  let running_jobs =
+    let help = "Number of jobs currently running" in
+    Gauge.v ~help ~namespace ~subsystem "running_jobs"
+end
 
-let solve_of_custom c =
-  let open Solver_service_api.Raw in
-  let payload = Cluster_api.Custom.payload c in
-  let r = Reader.of_pointer payload in
-  let request = Reader.Solver.Solve.Params.request_get r in
-  Solver_service_api.Worker.Solve_request.of_yojson
-  @@ Yojson.Safe.from_string request
+type t = {
+  solver : Solver_service.Solver.t;
+  name : string;
+  registration_service : Cluster_api.Raw.Client.Registration.t Sturdy_ref.t;
+  capacity : int;
+  mutable in_use : int; (* Number of active builds *)
+  cond : unit Lwt_condition.t;
+  (* Fires when a build finishes (or switch turned off) *)
+}
 
-let cluster_worker_log log =
-  let module L = Solver_service_api.Raw.Service.Log in
-  L.local
-  @@ object
-       inherit L.service
+let metrics = function
+  | `Agent ->
+    let open Lwt.Infix in
+    let content_type = "text/plain; version=0.0.4; charset=utf-8" in
+    Prometheus.CollectorRegistry.(collect default) >>= fun data ->
+    Lwt_result.return
+      ( content_type,
+        Fmt.to_to_string Prometheus_app.TextFormat_0_0_4.output data )
+  | `Host ->
+    failwith "No host metrics from solver service"
 
-       method write_impl params release_param_caps =
-         let open L.Write in
-         release_param_caps ();
-         let msg = Params.msg_get params in
-         Log_data.write log msg;
-         Capnp_rpc_lwt.Service.(return (Response.create_empty ()))
-     end
+let build ~cancelled ~log t descr =
+  let module R = Cluster_api.Raw.Reader.JobDescr in
+  match Cluster_api.Submission.get_action descr with
+  | Custom_build c ->
+    Log.info (fun f ->
+        f "Got request to build a job of kind \"%s\""
+          (Cluster_api.Custom.kind c));
+    (* Oddly, the protocol has us report cancellation and errors as "successful" jobs with the error inside! *)
+    let output = Custom.solve ~cancelled ~solver:t.solver ~log c in
+    Log_data.write log "Job succeeded\n";
+    (Ok output, "ok")
+  | _ ->
```

Comment:
```suggestion
  | Obuilder_build _ | Docker_build _ ->
```

Reply:
Why? We don't care if OCluster adds more job types here.

Reply:
Personal preference to use exhaustive pattern matching and get the warning IF more job types get added.